src/plugins/push/pushactivation.ts

```diff
@@ -0,0 +1,670 @@
+      this.callCustomRegisterer(localDevice, false);
+    } else {
+      const client = this.client;
+      const format = client.options.useBinaryProtocol
+          ? this.client.Utils.Format.msgpack
+          : this.client.Utils.Format.json,
+        body = client.rest.DeviceDetails.fromLocalDevice(localDevice),
+        headers = this.client.Defaults.defaultPostHeaders(this.client.options, { format }),
+        params = {};
+
+      if (client.options.headers) {
+        this.client.Utils.mixin(headers, client.options.headers);
+      }
+
+      if (client.options.pushFullWait) {
+        this.client.Utils.mixin(params, { fullWait: 'true' });
+      }
+
+      const requestBody = this.client.Utils.encodeBody(body, client._MsgPack, format);
+      const authDetails = localDevice.getAuthDetails(client, headers, params);
+      try {
+        const response = await this.client.rest.Resource.patch(
+          client,
+          '/push/deviceRegistrations',
+          requestBody,
+          authDetails.headers,
+          authDetails.params,
+          format,
+          true,
+        );
+        this.handleEvent(new GotDeviceRegistration(response.body as DeviceRegistration));
+      } catch (err) {
+        this.handleEvent(new GettingDeviceRegistrationFailed(err as ErrorInfo));
+      }
+    }
+  }
+
+  async deregister() {
+    const device = this.client.device as LocalDevice;
+    if (this.deregisterCallback) {
+      this.callCustomDeregisterer(device);
+    } else {
+      const rest = this.client;
+      const format = rest.options.useBinaryProtocol ? this.client.Utils.Format.msgpack : this.client.Utils.Format.json,
+        headers = this.client.Defaults.defaultPostHeaders(rest.options, { format }),
+        params = { deviceId: device.id };
+
+      if (rest.options.headers) this.client.Utils.mixin(headers, rest.options.headers);
+
+      if (rest.options.pushFullWait) this.client.Utils.mixin(params, { fullWait: 'true' });
+
+      try {
+        await this.client.rest.Resource.delete(rest, '/push/deviceRegistrations', headers, params, format, true);
+        this.handleEvent(new Deregistered());
+      } catch (err) {
+        this.handleEvent(new DeregistrationFailed(err as ErrorInfo));
+      }
+    }
+  }
+
+  callActivatedCallback(err: ErrorInfo | null) {
+    this.activatedCallback?.(err);
+    delete this.activatedCallback;
+  }
+
+  callDeactivatedCallback(err: ErrorInfo | null) {
+    this.deactivatedCallback?.(err);
+    delete this.deactivatedCallback;
+  }
+
+  handleEvent(event: ActivationEvent) {
+    if (this.handling) {
+      this.client.Platform.Config.nextTick(() => {
+        this.handleEvent(event);
+      });
+      return;
+    }
+
+    this.handling = true;
+    this.client.Logger.logAction(
+      this.client.logger,
+      this.client.Logger.LOG_MAJOR,
+      'Push.ActivationStateMachine.handleEvent()',
+      'handling event ' + event.constructor.name + ' from ' + this.current.name,
```

Comment:
`event.constructor.name` - need to do the same thing you did for state classes in https://github.com/ably/ably-js/pull/1775#discussion_r1639386108, otherwise names will be scrambled in minified code. Have an explicit `name` prop set in a constructor

Reply:
> Do we even use formFactor somewhere? Is this part of the realtime implementation and it's a minor metadata for some analytics?

No, it's not used anywhere - it's just informative for anyone who looks at the registration, eg via the push admin API